src/rust.rs

```diff
@@ -1257,3 +1258,73 @@ macro_rules! new_jsjitinfo_bitfield_1 {
             (($slotIndex as u32) << 22u32)
     }
 }
+
+pub struct CapturedJSStack<'a> {
+    cx: *mut JSContext,
+    stack: RootedGuard<'a, *mut JSObject>,
+}
+
+impl<'a> CapturedJSStack<'a> {
+    pub fn new(cx: *mut JSContext, mut guard: RootedGuard<'a, *mut JSObject>, max_frame_count: u32) -> Option<Self> { // or maybe another name, for example "capture"?
+        let obj_handle = guard.handle_mut(); // if nullptr, return None
+                                             // how to check if handle is not nullptr?
+                                             // how to handle such errors in a nice way?
+        unsafe {
+            if !CaptureCurrentStack(cx, obj_handle, max_frame_count) {
+                return None;
+            }
+        }
+
+        Some(CapturedJSStack {
+            cx: cx,
+            stack: guard,
+        })
+    }
+
+    pub fn as_string(&self, indent: usize) -> Option<String> {
+        // TODO where errors can occur? some example of non trivial error handling?
+        // TODO how to write tests for this code? Including the problem of obtaining some stack trace
+        // do we want to do the actual logic of conversion in this method or maybe in constructor and just return cached result?
+        unsafe { // since almost everything here is unsafe, is it okay to wrap it like this?
+            let stack_handle = self.stack.handle();
+            rooted!(in(self.cx) let mut js_string = JS_NewStringCopyZ(self.cx, ptr::null()));
+            let string_handle = js_string.handle_mut(); // as above -- how to check if is not nullptr?
+
+            if !BuildStackString(self.cx, stack_handle, string_handle, indent) {
+                return None;
+            }
+
+            // wanted somehow to just get the utf-16 string and use String::from_utf16_lossy
+            // but JS_GetTwoByteStringCharsAndLength requires some object of strange type (AutoCheckCannotGC)
+            let str_contents = JS_EncodeStringToUTF8(self.cx, string_handle.handle());
+            if str_contents.is_null() {
+                return None;
+            }
+
+            let mut size = 0 as usize;
+            while *str_contents.offset(size as isize) != 0 { // what is some better way of calculating length on raw pointers?
+                size += 1;
+            }
+
+            let str_slice = slice::from_raw_parts(str_contents as *const u8, size);
+            let result = String::from_utf8_lossy(str_slice).to_string();
+            JS_free(self.cx, str_contents as *mut ::std::os::raw::c_void);
+            Some(result)
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! capture_stack {
+    ($cx:expr, $max_frame_count:expr) => {
+        rooted!(in($cx) let mut __obj = JS_NewPlainObject($cx));
```

Comment:
We should be able to do `rooted!(in($cx) let mut __obj);` instead.

Reply:
Right now, we can't. Please take a look at the definition of the macro: https://github.com/servo/rust-mozjs/blob/master/src/rust.rs#L454-L464.
Should I expand it and if yes, would you prefer to do it in this PR or separate one? Another problem is that `JS_NewPlainObject` returns `*mut JSObject` and `JS_NewCopyStringZ`, which I used somewhere else, returns `*mut JSString`. So, what should be the type of the new binding created within the macro?

Reply:
Oh, my mistake. It should be possible to use a null pointer instead, though.

Reply:
So, if some function expects a mutable handle for an output argument, we can just pass a handle to a null pointer? Did I get it right?

Reply:
Yes. The mutable handle allows the function to initialize it with a meaningful value.